src/hash_to_curve/map_g2.rs

```diff
@@ -0,0 +1,900 @@
+        0x7bcf_a7a2_5aa3_0fda,
+        0xdc17_dec1_2a92_7e7c,
+        0x2f08_8dd8_6b4e_bef1,
+        0xd1ca_2087_da74_d4a7,
+        0x2da2_5966_96ce_bc1d,
+        0x0e2b_7eed_bbfd_87d2,
+    ]),
+    c1: Fp::from_raw_unchecked([
+        0x7bcf_a7a2_5aa3_0fda,
+        0xdc17_dec1_2a92_7e7c,
+        0x2f08_8dd8_6b4e_bef1,
+        0xd1ca_2087_da74_d4a7,
+        0x2da2_5966_96ce_bc1d,
+        0x0e2b_7eed_bbfd_87d2,
+    ]),
+};
+
+impl HashToField for Fp2 {
+    // ceil(log2(p)) = 381, m = 2, k = 128.
+    type InputLength = U128;
+
+    fn from_okm(okm: &GenericArray<u8, U128>) -> Fp2 {
+        let c0 = <Fp as HashToField>::from_okm(GenericArray::<u8, U64>::from_slice(&okm[..64]));
+        let c1 = <Fp as HashToField>::from_okm(GenericArray::<u8, U64>::from_slice(&okm[64..]));
+        Fp2 { c0, c1 }
+    }
+}
+
+impl Sgn0 for Fp2 {
+    fn sgn0(&self) -> Choice {
+        let sign_0 = self.c0.sgn0();
+        let zero_0 = self.c0.is_zero();
+        let sign_1 = self.c1.sgn0();
+        sign_0 | (zero_0 & sign_1)
+    }
+}
+
+/// Maps from an [`Fp2]` element to a point on iso-G2.
+fn map_to_curve_simple_swu(u: &Fp2) -> G2Projective {
+    let usq = u.square();
+    let xi_usq = SSWU_XI * usq;
+    let xisq_u4 = xi_usq.square();
+    let nd_common = xisq_u4 + xi_usq; // XI^2 * u^4 + XI * u^2
+    let x_den = SSWU_ELLP_A * Fp2::conditional_select(&(-nd_common), &SSWU_XI, nd_common.is_zero());
+    let x0_num = SSWU_ELLP_B * (Fp2::one() + nd_common); // B * (1 + (XI^2 * u^4 + XI * u^2))
+
+    // compute g(x0(u))
+    let x_densq = x_den.square();
+    let gx_den = x_densq * x_den;
+    // x0_num^3 + A * x0_num * x_den^2 + B * x_den^3
+    let gx0_num = (x0_num.square() + SSWU_ELLP_A * x_densq) * x0_num + SSWU_ELLP_B * gx_den;
+
+    // compute g(x0(u)) ^ ((p^2 - 9) // 16)
+    let sqrt_candidate = {
+        let vsq = gx_den.square(); // v^2
+        let v_3 = vsq * gx_den; // v^3
+        let v_4 = vsq.square(); // v^4
+        let uv_7 = gx0_num * v_3 * v_4; // u v^7
+        let uv_15 = uv_7 * v_4.square(); // u v^15
+        uv_7 * chain_p2m9div16(&uv_15) // u v^7 (u v^15) ^ ((p^2 - 9) // 16)
+    };
+
+    // set y = sqrt_candidate * Fp2::one(), check candidate against other roots of unity
+    let mut y = sqrt_candidate;
+    // check Fp2(0, 1)
+    let tmp = Fp2 {
+        c0: -sqrt_candidate.c1,
+        c1: sqrt_candidate.c0,
+    };
+    y.conditional_assign(&tmp, (tmp.square() * gx_den).ct_eq(&gx0_num));
+    // check Fp2(RV1, RV1)
+    let tmp = sqrt_candidate * SSWU_RV1;
+    y.conditional_assign(&tmp, (tmp.square() * gx_den).ct_eq(&gx0_num));
+    // check Fp2(RV1, -RV1)
+    let tmp = Fp2 {
+        c0: tmp.c1,
+        c1: -tmp.c0,
+    };
+    y.conditional_assign(&tmp, (tmp.square() * gx_den).ct_eq(&gx0_num));
+
+    // compute g(x1(u)) = g(x0(u)) * XI^3 * u^6
+    let gx1_num = gx0_num * xi_usq * xisq_u4;
+    // compute g(x1(u)) * u^3
+    let sqrt_candidate = sqrt_candidate * usq * u;
```

Comment:
This differs from Appendix G.2.3 of draft 11, which uses `y * usq * u` in steps 44-45. That has an effect on how steps 46-67 are implemented.